ballista/core/src/physical_optimizer/task_group.rs

```diff
@@ -166,18 +136,29 @@ impl PhysicalOptimizerRule for OptimizeTaskGroup {
     }
 }
 
+/// Return whether we should insert a `CoalesceTasksExec` above this node.
+fn insert_coalesce(plan: &dyn ExecutionPlan) -> bool {
```

Comment:
I wonder if we're actually 100% correct after this PR. 
If the upstream stage would depend on e.g. distribution (hash partitioning) or sorting being preserved, we could undo the sorting / partitioning by doing the coalescing and writing to the same shuffle files.

Maybe we should an issue which does part of the planning / optimization *before* the required distribution and sorting are satisfied in order to create a correct plan.